### PR TITLE
Clean up the code for Maps::Tiles

### PR DESCRIPTION
**Pull request by ihhub**
- remove **_hasObjectAnimation** and **_isMarkedAsRoad** members from Maps::Tiles and Maps::TilesAddon since they were not used anywhere in the code
- remove the usage of bit flags for the above variables. We shouldn't read them
- rename multiple variable for structures for better understanding what they are for
- make the logic of loading data from MP2 format clear
- simplify the logic for **_isTileMarkedAsRoad** (relates to #7772)
- if an object is set we shouldn't check image index as it is not necessary
- remove weird logic for **layerType**

This pull request should be merged before #7772 in order to have much cleaner code for the future maintenance. There are more changes to be done but it is better to do them separately to avoid too complex changes in one shot.

This pull request request needs thoughtful testing to make sure that it does not break passability rules and does not introduce any rendering issues. How to test this pull request:
- load multiple maps and save them in debug mode within View World (set VIEWWORLD_DEBUG_ZOOM_LEVEL)
- compare them with master branch for rendering and passability rules
- load multiple save files and check that rendering and passabilities are correct

### CODE REVIEW

**Comment by ihhub** [comment]
Hi @Branikolog , please priotirize testing this pull request over #7772 as the last one depends on these changes.

File: src/fheroes2/maps/maps_tiles.cpp [hunk 1]
```diff
@@ -419,65 +419,41 @@ namespace
...
-    case MP2::OBJ_ICN_TYPE_OBJNTOWN:
-        if ( 13 == _imageIndex || 29 == _imageIndex || 45 == _imageIndex || 61 == _imageIndex || 77 == _imageIndex || 93 == _imageIndex || 109 == _imageIndex
-             || 125 == _imageIndex || 141 == _imageIndex || 157 == _imageIndex || 173 == _imageIndex || 189 == _imageIndex )
-            return true;
-        break;
-
-    // Random castle or town gate.
-    case MP2::OBJ_ICN_TYPE_OBJNTWRD:
-        return ( _imageIndex == 13 || _imageIndex == 29 );
+    bool isSpriteRoad( const MP2::ObjectIcnType objectIcnType, const uint8_t imageIndex )
```
Comment (ihhub) [inline src/fheroes2/maps/maps_tiles.cpp:423]: Hi @Districh-ru , could you please verify this function very carefully. Essentially, all image indices here are used as main road indices. I could be wrong as I just modified the old function.

Comment (Districh-ru) [reply src/fheroes2/maps/maps_tiles.cpp:423]: I checked the image indexes (thanks to XnView "create contact sheet") and all is OK.

**Review by Districh-ru — CHANGES_REQUESTED** [review]
Hi, @ihhub.
While testing I found an issue with save files: try to save any game and load it to fail the assertion:
`Assertion failed: 0, file G:\Games\heroes2\git\src\fheroes2\maps\maps_tiles.cpp, line 2035`
Save file: [AUTOSAVE.zip](https://github.com/ihhub/fheroes2/files/12705879/AUTOSAVE.zip)
Loading save files made in master branch works fine.
Could you please check this issue when you have time.

**Review by oleg-derevenetz — APPROVED** [review]

File: src/fheroes2/maps/maps_tiles.cpp [hunk 2]
```diff
@@ -1993,23 +1979,24 @@ StreamBase & Maps::operator>>( StreamBase & msg, Tiles & tile )
...
+            msg >> objectIcnType;
 
-        tile._objectIcnType = static_cast<MP2::ObjectIcnType>( objectIcnType );
+            tile._objectIcnType = static_cast<MP2::ObjectIcnType>( objectIcnType );
 
-        msg >> tile._hasObjectAnimation >> tile._isMarkedAsRoad;
+            bool temp;
+            msg >> temp >> temp;
+        }
     }
```
Comment (Districh-ru) [inline src/fheroes2/maps/maps_tiles.cpp:2001]: In the case if we load the `FORMAT_VERSION_1009_RELEASE` save version we should not skip loading the `_objectIcnType`:
```suggestion
    else {
        ObjectIcnTypeUnderlyingType objectIcnType = MP2::OBJ_ICN_TYPE_UNKNOWN;
        msg >> objectIcnType;

        tile._objectIcnType = static_cast<MP2::ObjectIcnType>( objectIcnType );
    }

```

File: src/fheroes2/maps/maps_tiles.cpp [hunk 3]
```diff
@@ -1917,42 +1898,47 @@ void Maps::Tiles::swap( TilesAddon & addon ) noexcept
...
+            msg >> objectIcnType;
 
-        ta._objectIcnType = static_cast<MP2::ObjectIcnType>( objectIcnType );
+            ta._objectIcnType = static_cast<MP2::ObjectIcnType>( objectIcnType );
 
-        msg >> ta._hasObjectAnimation >> ta._isMarkedAsRoad;
+            bool temp;
+            msg >> temp >> temp;
+        }
     }
```
Comment (Districh-ru) [inline src/fheroes2/maps/maps_tiles.cpp:1943]: In the case if we load the `FORMAT_VERSION_1009_RELEASE` save version we should not skip loading the `_objectIcnType`:
```suggestion
    else {
        ObjectIcnTypeUnderlyingType objectIcnType = MP2::OBJ_ICN_TYPE_UNKNOWN;
        msg >> objectIcnType;

        ta._objectIcnType = static_cast<MP2::ObjectIcnType>( objectIcnType );
    }

```

**Review by Districh-ru — CHANGES_REQUESTED** [review]
Hi, @ihhub, I left three comments regarding save format issue I found earlier. Could you please check them in your spare time.
I compared the passability in some new maps and loaded saves with the master build and found no issues.

**Review by Districh-ru — APPROVED** [review]

**Review by Branikolog — APPROVED** [review]

**Pull request merged**